test/distributed/_shard/checkpoint/test_checkpoint.py

```diff
@@ -260,5 +283,227 @@ def test_create_key_handles_collision(self):
         key1 = _create_storage_key(keys, "foo")
         self.assertNotEqual(key0, key1)
 
+
+
+
+class TestStorageBase:
+    def __init__(
+        self,
+        fail_conf
+    ):
+        self.fail_conf = fail_conf
+        self.rank = 0 if not dist.is_initialized() else dist.get_rank()
+
+    def _get_ranks(self, name):
+        return self.fail_conf[name] if name in self.fail_conf else None
+
+    def _fail_rank(self, name):
+        ranks = self._get_ranks(name)
+        if ranks is not None and self.rank in ranks:
+            raise ValueError(f"rank fail {self.rank} for {name}")
+
+    def _fail_rank_async(self, name):
+        ranks = self._get_ranks(name)
+        fut = Future()
+        if ranks is not None and self.rank in ranks:
+            fut.set_exception(ValueError(f"async rank fail {self.rank} for {name}"))
+        else:
+            fut.set_result(None)
+        return fut
+
+
+class TestStorageWriter(TestStorageBase, StorageWriter):
```

Comment:
nit: The `Test` prefix here was a little confusing in the sense that I thought this is an actual unit test class but it is just a StorageWriter for testing purposes. Maybe rename this to something like `SimpleStorageWriter`.

Reply:
I renamed them to FaultyStorageWriter and FaultyStorageReader.